### PR TITLE
V1.4 fix inconsitent error messages for invalid index

### LINKED ISSUE #217: [PE-D] Bugs on `delete` command
Different invalid index numbers results in different message:

-------------
Labels: `severity.Low` `type.FunctionalityBug`
original: eltonyeh/ped#9

### LINKED ISSUE #223: [PE-D] Using index-based commands with index above INT_MAX_VALUE shows wrong error message
If the index does not refer to a student, then the above error message should be shown. 

However, when a large enough number is provided, the error message is "invalid command format" instead.

-------------
Labels: `severity.VeryLow` `type.FunctionalityBug`
original: jyrw/ped#7

### PR DESCRIPTION
**Pull request by nbvannhi**
Closes #223 
Closes #217 

`INDEX` error message summary
- If `INDEX` is specified but command is invalid,
_invalid command format_ error message will be shown with correct command format.
- If `INDEX` is unable to parse,
_invalid index_ error message will be shown (including both negative and exceeding `MAX_INT` cases).
- If `INDEX` exceeds no. students,
_invalid student index_ error message will be thrown with no. students.

Note: `edit 1 some_random_texts` will shown _invalid index_ error message instead of _invalid command format_ error message.

### CODE REVIEW

File: src/main/java/seedu/address/logic/commands/ShowCommand.java [hunk 1]
```diff
@@ -121,7 +121,7 @@ private CommandResult showStudentByIndex(Model model) throws CommandException {
 
         List<Student> students = model.getFilteredStudentList();
 
-        if (index.getZeroBased() >= students.size()) {
+        if (index.getOneBased() < 1 || index.getZeroBased() >= students.size()) {
```
Comment (jonas-chow) [inline src/main/java/seedu/address/logic/commands/ShowCommand.java]: The added clause will never be true. `ParseIndex` takes care of this case already.

File: src/main/java/seedu/address/logic/parser/ShowCommandParser.java [hunk 2]
```diff
@@ -47,7 +48,7 @@ public ShowCommand parseByIndex(ArgumentMultimap argMultimap, Path savePath) thr
         try {
             index = ParserUtil.parseIndex(argMultimap.getPreamble());
         } catch (ParseException pe) {
-            throw new ParseException(String.format(MESSAGE_INVALID_COMMAND_FORMAT, ShowCommand.MESSAGE_USAGE));
+            throw new ParseException(MESSAGE_INVALID_STUDENT_DISPLAYED_INDEX);
```
Comment (jonas-chow) [inline src/main/java/seedu/address/logic/parser/ShowCommandParser.java]: Perhaps we need a more descriptive error message than "The student index provided is invalid".
I suggest we split the error message into the following cases:

* Inputs that contain characters that are not digits: "Index is not a positive unsigned integer."
* Inputs that only contain digits
    * Those that `parseInt` properly: 
        * If 0: "Index is not a positive unsigned integer."
        * If greater than student list size: "Index is too large"
        * Else: Executes properly
    * Doesn't `parseInt`: Above `MAX_INT`: "Index is too large"

Would be nice if you could rewrite the `parseIndex` function in `ParserUtil.java` to implement these (except perhaps the greater than student list size one)

If it's in the `parseIndex` function, then it'd be general enough to also affect the other related commands like `edit` and `delete`.

**Comment by zhing22** [comment]
yes could you please help to fix all other commands that uses index? (namely show, delete, and edit)

also since you are working on it could you take over #217 too? its about error message for negative numbers and 0.

Thank you!! <3

**Review by jonas-chow — COMMENTED** [review]
I feel like the error message is too general here.
Things like `show a` and `show -1` will give the same error message, but the error message doesn't really tell the user how they should be changing their input to make it valid.

**Review by jonas-chow — CHANGES_REQUESTED** [review]
`show 1` doesn't work!!!

**Review by jonas-chow — APPROVED** [review]
There was an issue with show command not working with indices. I fixed it by negating a condition. Otherwise, lgtm.

**Comment by nbvannhi** [comment]
> 
> 
> `show 1` doesn't work!!!

Oh I fixed it later but forgot to push... Tks anw!

**Pull request merged**